src/realizations/catchment/Bmi_Cpp_Adapter.cpp

```diff
@@ -0,0 +1,235 @@
+#include <exception>
+#include <utility>
+
+#include "FileChecker.h"
+#include "Bmi_Cpp_Adapter.hpp"
+#include "boost/algorithm/string.hpp"
+
+using namespace models::bmi;
+
+Bmi_Cpp_Adapter::Bmi_Cpp_Adapter(const std::string& type_name, std::string library_file_path, std::string forcing_file_path,
+                             bool allow_exceed_end, bool has_fixed_time_step,
+                             const std::string& creator_func, const std::string& destroyer_func,
+                             utils::StreamHandler output)
+        : Bmi_Cpp_Adapter(type_name, std::move(library_file_path), "", std::move(forcing_file_path),
+                        allow_exceed_end, has_fixed_time_step, creator_func, destroyer_func, output) { }
+
+Bmi_Cpp_Adapter::Bmi_Cpp_Adapter(const std::string& type_name, std::string library_file_path, std::string bmi_init_config,
+                             std::string forcing_file_path, bool allow_exceed_end, bool has_fixed_time_step,
+                             std::string creator_func, std::string destroyer_func,
+                             utils::StreamHandler output)
+        : Bmi_Cpp_Adapter(type_name, std::move(library_file_path), std::move(bmi_init_config),
+                        std::move(forcing_file_path), allow_exceed_end, has_fixed_time_step,
+                        std::move(creator_func), std::move(destroyer_func), output, true) { }
+
+Bmi_Cpp_Adapter::Bmi_Cpp_Adapter(const std::string& type_name, std::string library_file_path, std::string bmi_init_config,
+                             std::string forcing_file_path, bool allow_exceed_end, bool has_fixed_time_step,
+                             std::string creator_func, std::string destroyer_func,
+                             utils::StreamHandler output, bool do_initialization)
+                             : AbstractCLibBmiAdapter<Cpp_Bmi>(type_name, library_file_path, std::move(bmi_init_config), std::move(forcing_file_path), allow_exceed_end,
+                             has_fixed_time_step, creator_func, output)
+                             //TODO: We are passing creator_func as registration_func because AbstractCLibBmiAdapter expects it to exist, but are not using it the same way...may be okay but we may want to generalize that assumption out!
```

Comment:
I'm inclined to keep naming them differently because the C++ common practice for what we're doing strongly implies memory management responsibility, while `register` does not...but if I name them the same to I get to skip the refactoring required by this `TODO`?

Reply:
`Bmi_Fortran_Adapter` also uses the `bmi_registration_func` parameter, which is probably why it was carried into the parent. So there's some utility in that being in the parent. This may just need a clever name and adjustment to the error messages to generalize it.

Reply:
Many ways to accomplish the same thing, but none of which we can offer string guarantees for.  Should definitely keep this on the back burner for when we get bored an want discuss the merits of `BMI::Finalize` requiring ALL memory management cleanup in the library vs having `BMI::Finalize` and `destroy` and how those interact with plugin/library registration.

Reply:
Concur. Yeah it's interesting... or, weird... looking at [`bmi_heat.c`](https://github.com/csdms/bmi-example-c/blob/master/heat/bmi_heat.c#L133) it's as if `Finalize()` is supposed to _set up the model for a new run_...is that the intent??